File: quickreserve/server/src/services/twogis.js
const TWO_GIS_BASE = "https://catalog.api.2gis.com/3.0/items";

async function importFrom2GIS({ apiKey, city = "almaty", q = "service" }) {
  if (!apiKey) {
    return [];
  }

  const url = new URL(TWO_GIS_BASE);
  url.searchParams.set("q", q);
  url.searchParams.set("city", city);
  url.searchParams.set("fields", "items.point,items.address_name,items.schedule");
  url.searchParams.set("key", apiKey);
  url.searchParams.set("page_size", "20");

  const response = await fetch(url.toString());
  if (!response.ok) {
    throw new Error(`2GIS request failed: ${response.status}`);
  }

  const payload = await response.json();
  const items = payload?.result?.items || [];

  return items
    .filter((item) => item?.point?.lat && item?.point?.lon)
    .map((item, index) => ({
      id: `2gis-${item.id || index}`,
      name: item.name,
      category: item.purpose_name || "service",
      address: item.address_name || "",
      phone: item?.contact_groups?.[0]?.contacts?.[0]?.value || "",
      rating: 4.0,
      priceLevel: 2,
      lat: item.point.lat,
      lng: item.point.lon,
      workingHours: item?.schedule?.working_hours_text || "",
      slots: [],
    }));
}

module.exports = { importFrom2GIS };

File: quickreserve/backend_py/app/services/twogis.py
import json
from typing import Any

import httpx

from app.core.config import settings
from app.services.redis_client import redis_call


async def _allow_twogis_call() -> bool:
    key = "ratelimit:2gis:minute"
    current = await redis_call("incr", key)
    if current == 1:
        await redis_call("expire", key, 60)
    return current <= settings.TWOGIS_RPM_LIMIT


async def search_nearby(city: str, lat: float, lng: float, radius_km: float, category: str | None) -> list[dict[str, Any]]:
    if not settings.TWOGIS_API_KEY:
        return []

    cache_key = f"2gis:nearby:{city}:{lat:.4f}:{lng:.4f}:{radius_km}:{category or 'all'}"
    cached = await redis_call("get", cache_key)
    if cached:
        return json.loads(cached)

    if not await _allow_twogis_call():
        return []

    params = {
        "q": category or "service",
        "city": city,
        "point": f"{lng},{lat}",
        "radius": int(radius_km * 1000),
        "fields": "items.point,items.address_name,items.reviews,items.rubrics,items.schedule",
        "key": settings.TWOGIS_API_KEY,
        "page_size": 50,
    }

    async with httpx.AsyncClient(timeout=10) as client:
        response = await client.get(settings.TWOGIS_BASE_URL, params=params)
        response.raise_for_status()
        payload = response.json()

    items = payload.get("result", {}).get("items", [])
    normalized = []
    for item in items:
        point = item.get("point") or {}
        if not point.get("lat") or not point.get("lon"):
            continue
        normalized.append(
            {
                "gis_id": str(item.get("id", "")),
                "name": item.get("name", "Unknown"),
                "address": item.get("address_name"),
                "lat": point["lat"],
                "lng": point["lon"],
                "rating": ((item.get("reviews") or {}).get("general_rating")),
            }
        )

    await redis_call("setex", cache_key, 120, json.dumps(normalized))
    return normalized
